[Build Speed] Reduce includes of HTMLMediaElement.h and HTMLVideoElement.h

jernoble opened the pull request:
#### 5033ebb318e52dbc35184d414ba0a1947803fca0
<pre>
[Build Speed] Reduce includes of HTMLMediaElement.h and HTMLVideoElement.h
<a href="https://rdar.apple.com/163794224">rdar://163794224</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=301755">https://bugs.webkit.org/show_bug.cgi?id=301755</a>

Reviewed by Anne van Kesteren.

Reduce the number of times HTMLMediaElement.h and HTMLVideoElement.h are included during a build,
mostly by removing those includes from other headers. Sometimes the include can be replaced with a
simple forward declaration, sometimes it can be replaced by redeclaring an enum type, and sometimes
the include was necessary for an inlined function. In the latter cases, the inlined function was
moved into the implementation file when there was no benefit from inlining the function, or
an -Inlines.h header was made and the inlined functions were moved there.  In one case, Document.h
was including HTMLMediaElement.h because the bindings specified a function would return an Element,
and the function in Document.h was returning a HTMLVideoElement, forcing the caller to cast. Since
the caller was a generated source, there was no way to move the include there; instead an explicit
mixin class DocumentOrShadowRootPictureInPicture.

Before these changes, HTMLMediaElement.h and HTMLVideoElement.h were included 102 and 65 times
respectively, taking 58s of CPU time to parse during the WebCore build.

After these changes, HTMLMediaElement.h and HTMLVideoElement.h were included 75 and 53 times
respectively, taking 44s of CPU time to parse during the WebCore build.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.h:
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRoot+PictureInPicture.idl:
* Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRootPictureInPicture.cpp: Copied from Source/WebCore/Modules/pictureinpicture/DocumentPictureInPicture.h.
(WebCore::DocumentOrShadowRootFullscreen::pictureInPictureElement):
* Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRootPictureInPicture.h: Copied from Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRoot+PictureInPicture.idl.
* Source/WebCore/Modules/pictureinpicture/DocumentPictureInPicture.cpp:
(WebCore::DocumentPictureInPicture::exitPictureInPicture):
* Source/WebCore/Modules/pictureinpicture/DocumentPictureInPicture.h:
* Source/WebCore/Modules/webaudio/MediaElementAudioSourceNode.cpp:
* Source/WebCore/Modules/webaudio/MediaElementAudioSourceNode.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::isMediaObject const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::pictureInPictureElement const):
* Source/WebCore/dom/DocumentFullscreen.cpp:
* Source/WebCore/dom/DocumentFullscreen.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::pictureInPictureElement const): Deleted.
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/html/HTMLElement.cpp:
* Source/WebCore/html/HTMLMediaElement.cpp:
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLMediaElementEnums.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/html/MediaElementSession.cpp:
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
* Source/WebCore/page/DeprecatedGlobalSettings.cpp:
* Source/WebCore/page/DiagnosticLoggingKeys.cpp:
* Source/WebCore/rendering/RenderMedia.cpp:
* Source/WebCore/rendering/RenderMedia.h:
(WebCore::RenderMedia::mediaElement const): Deleted.
(WebCore::RenderMedia::protectedMediaElement const): Deleted.
(WebCore::HTMLMediaElement::hasRenderer const): Deleted.
(WebCore::HTMLMediaElement::renderer const): Deleted.
* Source/WebCore/rendering/RenderMediaInlines.h: Copied from Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRoot+PictureInPicture.idl.
(WebCore::HTMLMediaElement::hasRenderer const):
(WebCore::HTMLMediaElement::renderer const):
(WebCore::RenderMedia::mediaElement const):
(WebCore::RenderMedia::protectedMediaElement const):
* Source/WebCore/rendering/RenderVideo.cpp:
* Source/WebCore/rendering/RenderVideo.h:
(WebCore::HTMLVideoElement::renderer const): Deleted.
* Source/WebCore/rendering/RenderVideoInlines.h: Copied from Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRoot+PictureInPicture.idl.
(WebCore::HTMLVideoElement::renderer const):
* Source/WebCore/style/UserAgentStyle.cpp:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/ShareableBitmapUtilities.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:

Canonical link: <a href="https://commits.webkit.org/302498@main">https://commits.webkit.org/302498@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/208eb828ccbc7c2b0a528a55aa9d7d20e8b77b80

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129273 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1529 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40110 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136649 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80666 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6a87035f-8162-457c-a203-d64c69ac995a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131144 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1461 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1406 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98443 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66352 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4c22636a-000d-4432-8bf2-341c4e62352b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132220 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1141 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115798 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79094 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1b96f02d-d684-465c-b525-40227e6f78f6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1075 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33917 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79927 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/109520 "Build was cancelled. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; 1 api test failed or timed out; re-run-api-tests (cancelled)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34412 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139122 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1320 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1270 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106970 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1368 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112134 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106809 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27197 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1090 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30653 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53943 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1392 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64748 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1214 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/1250 "Build was cancelled. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1314 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -11410,7 +11406,7 @@ void Document::dispatchSystemPreviewActionEvent(const SystemPreviewInfo& systemP
 
 #if ENABLE(PICTURE_IN_PICTURE_API)
-HTMLVideoElement* Document::pictureInPictureElement() const
+Element* Document::pictureInPictureElement() const
```
This seems a bit weird given that m_pictureInPictureElement still holds an HTMLVideoElement. And the setter below also still takes an HTMLVideoElement. Perhaps HTMLVideoElement ends up being included anyway due to the way we build?

annevk replied:
(Also, if this is somehow correct, then we know it's an HTMLVideoElement or null and we shouldn't use dynamicDowncast in that other file, we should just use downcast.)

jernoble replied:
This method implements a binding:

```
partial interface mixin DocumentOrShadowRoot {
  readonly attribute Element? pictureInPictureElement;
};
```

Which is defined to return an Element. The fact we only let a `HTMLVideoElement` become a picture-in-picture element in why it's stored as a `Ref<HTMLVideoElement>` and why the setter takes a `HTMLVideoElement`. 

This could be addressed with a `[ImplementedBy=pictureInPictureElementForBindings]`, but then it'd have to be changed in both the ShadowRoot and Document.

annevk reviewed (approved):
Commit message should note the functional change and arguably we should have a test for that as well.

jernoble committed:
[Build Speed] Reduce includes of HTMLMediaElement.h and HTMLVideoElement.h
rdar://163794224
https://bugs.webkit.org/show_bug.cgi?id=301755

Reviewed by Anne van Kesteren.

Reduce the number of times HTMLMediaElement.h and HTMLVideoElement.h are included during a build,
mostly by removing those includes from other headers. Sometimes the include can be replaced with a
simple forward declaration, sometimes it can be replaced by redeclaring an enum type, and sometimes
the include was necessary for an inlined function. In the latter cases, the inlined function was
moved into the implementation file when there was no benefit from inlining the function, or
an -Inlines.h header was made and the inlined functions were moved there.  In one case, Document.h
was including HTMLMediaElement.h because the bindings specified a function would return an Element,
and the function in Document.h was returning a HTMLVideoElement, forcing the caller to cast. Since
the caller was a generated source, there was no way to move the include there; instead an explicit
mixin class DocumentOrShadowRootPictureInPicture.

Before these changes, HTMLMediaElement.h and HTMLVideoElement.h were included 102 and 65 times
respectively, taking 58s of CPU time to parse during the WebCore build.

After these changes, HTMLMediaElement.h and HTMLVideoElement.h were included 75 and 53 times
respectively, taking 44s of CPU time to parse during the WebCore build.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.h:
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRoot+PictureInPicture.idl:
* Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRootPictureInPicture.cpp: Copied from Source/WebCore/Modules/pictureinpicture/DocumentPictureInPicture.h.
(WebCore::DocumentOrShadowRootFullscreen::pictureInPictureElement):
* Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRootPictureInPicture.h: Copied from Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRoot+PictureInPicture.idl.
* Source/WebCore/Modules/pictureinpicture/DocumentPictureInPicture.cpp:
(WebCore::DocumentPictureInPicture::exitPictureInPicture):
* Source/WebCore/Modules/pictureinpicture/DocumentPictureInPicture.h:
* Source/WebCore/Modules/webaudio/MediaElementAudioSourceNode.cpp:
* Source/WebCore/Modules/webaudio/MediaElementAudioSourceNode.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::isMediaObject const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::pictureInPictureElement const):
* Source/WebCore/dom/DocumentFullscreen.cpp:
* Source/WebCore/dom/DocumentFullscreen.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::pictureInPictureElement const): Deleted.
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/html/HTMLElement.cpp:
* Source/WebCore/html/HTMLMediaElement.cpp:
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLMediaElementEnums.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/html/MediaElementSession.cpp:
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
* Source/WebCore/page/DeprecatedGlobalSettings.cpp:
* Source/WebCore/page/DiagnosticLoggingKeys.cpp:
* Source/WebCore/rendering/RenderMedia.cpp:
* Source/WebCore/rendering/RenderMedia.h:
(WebCore::RenderMedia::mediaElement const): Deleted.
(WebCore::RenderMedia::protectedMediaElement const): Deleted.
(WebCore::HTMLMediaElement::hasRenderer const): Deleted.
(WebCore::HTMLMediaElement::renderer const): Deleted.
* Source/WebCore/rendering/RenderMediaInlines.h: Copied from Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRoot+PictureInPicture.idl.
(WebCore::HTMLMediaElement::hasRenderer const):
(WebCore::HTMLMediaElement::renderer const):
(WebCore::RenderMedia::mediaElement const):
(WebCore::RenderMedia::protectedMediaElement const):
* Source/WebCore/rendering/RenderVideo.cpp:
* Source/WebCore/rendering/RenderVideo.h:
(WebCore::HTMLVideoElement::renderer const): Deleted.
* Source/WebCore/rendering/RenderVideoInlines.h: Copied from Source/WebCore/Modules/pictureinpicture/DocumentOrShadowRoot+PictureInPicture.idl.
(WebCore::HTMLVideoElement::renderer const):
* Source/WebCore/style/UserAgentStyle.cpp:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/ShareableBitmapUtilities.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:

Canonical link: https://commits.webkit.org/302498@main
